Move tools/ci_* projects to airbyte-ci, update to use Poetry, bump to python 3.10

bnchrch opened the pull request:
## What

1. moves tools/ci_* projects to airbyte-ci
2. Updates them to use poetry/pipx
3. Updates these packages to python 3.10
4. Improves the installation story
5. Moves pipelines to its own package

Now to install airbyte-ci all you have to run is 

```
pipx install airbyte-ci/connectors/pipelines
```
## Notes for reviewer

@alafanechere I would love to get your take on this being the `airbyte-ci` tzar.

- Is this a good idea?
- Is this the direction you had wanted to go in?
- What do you think about using pipx to isolate the install?
- The folders `ci` and `connectors` at the top of `airbyte-ci` still feel awkward is there a better way?


## Left TODO
- [x] Update `ci_code_validator`
- [x] Update GHA
- [x] Update docs
- [x] Update Dagger build 
- [x] Test that everything still works

closes #25170 
closes #27865

bnchrch commented on airbyte-ci/ci/ci_credentials/README.md:
Pipelines/airbyte-ci are now their own package

bnchrch commented on airbyte-ci/ci/ci_credentials/README.md:
Pipelines/airbyte-ci are now their own package

bnchrch commented on airbyte-ci/ci/ci_pipelines/README.md:
Pipelines/airbyte-ci are now their own package

bnchrch commented on airbyte-ci/ci/ci_pipelines/README.md:
````diff
@@ -1,29 +1,43 @@
+Once pipx is installed, navigate to the root directory of the project, then run:
+
+```bash
+pipx install airbyte-ci/ci/ci_pipelines/
````
The installation is one line

bnchrch commented on airbyte-ci/ci/ci_credentials/pyproject.toml:
```diff
@@ -0,0 +1,29 @@
+packages = [{ include = "ci_credentials" }]
+
+[tool.poetry.dependencies]
+python = "^3.10"
```
Python 3.10 all around!!!

alafanechere replied:
ci_credentials is installed on GHA workflows that are running `3.9`. Can we keep compatibility of `ci_credentials` with `3.9` while suggesting use of `3.10` though `.python-version` files?

bnchrch replied:
I believe we can!

pedroslopez replied:
Is there a reason to keep the version diff between local and ci? Could this be another source of behavior mismatches?

pedroslopez replied:
nvm, seems like we bumped most things as we went along, except `.github/workflows/connector-performance-command.yml` for some reason?

bnchrch replied:
Oh I think that was a miss on my part.

I ended up going with 3.10 everywhere as it seems that there were no issues encountered while running these tools as 3.10

bnchrch commented on airbyte-ci/ci/ci_credentials/pyproject.toml:
```diff
@@ -0,0 +1,29 @@
+requests = "^2.28.2"
+click = "^8.1.3"
+pyyaml = "^6.0"
+ci_common_utils = { path = "../ci_common_utils", develop = true }
```
If we go ahead with this we will need to update the local dependency resolution right?

alafanechere commented:
> Is this a good idea?

Yes it is! 

> Is this the direction you had wanted to go in?

Totally

> What do you think about using pipx to isolate the install?

I'm not against it but I'm not familiar with it. Is it providing dependency isolation while not requiring the use of a python virtualenv? I'm 👍 , but if poetry has a builtin way to achieve the same I'd prefer to stick to poetry (I'm not too familiar with it too..)

> The folders ci and connectors at the top of airbyte-ci still feel awkward is there a better way?

So in this `ci_connector_ops`  /  `ci_credentials` / `ci_common_utils` thing we have:
* QA checks
* `airbyte-ci` pipelines
* `ci_credentials` to download GSm secrets
* `ci_common_utils` which only has a `GoogleAPI` client class... Which I believe is only used by `ci_credentials`.

My suggestion would be to: 
* Merge `ci_credentials` and `ci_common_utils` into a single package : `airbyte-ci/connectors/credentials` ?
 * I'd love to have a package we can use as a lib that would have the `Connector` utility class. `airbyte-ci/connectors/utils` ? It could be imported in `pipelines` and `qa_checks` for instance.
 * The QA checks should be a light package in `airbyte-ci/connectors/qa_checks`

alafanechere commented:
Allright, let's use pipx!
And let's refactor `ci_common_utils and ci_credentials` in a different PR!

> What I can do is rename these to remove ci and place them all under airbyte-ci/connectors/

👍 

> I'd love to have a package we can use as a lib that would have the Connector utility class. airbyte-ci/connectors/utils ? It could be imported in pipelines and qa_checks for instance.

Do you have an opinion about this?

bnchrch commented:
> > I'd love to have a package we can use as a lib that would have the Connector utility class. airbyte-ci/connectors/utils ? It could be imported in pipelines and qa_checks for instance.
> 
> Do you have an opinion about this?

I LOVE this. Thats my opinion. I think were in a great spot to do that. Also it would be useful in the metadata_service

bnchrch commented:
Pre-Publish that ran successfully: https://github.com/airbytehq/airbyte/actions/runs/5651515543/job/15309740249

bnchrch commented:
**GHA Connector Tests**
source-fresh-desk: https://github.com/airbytehq/airbyte/actions/runs/5651586425
source-mysql: https://github.com/airbytehq/airbyte/actions/runs/5651609514
destination-postgres: https://github.com/airbytehq/airbyte/actions/runs/5651597479

octavia-squidington-iii commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/26884cd0dbe194727b5982e7b67c96e7b60573a6/airbyte-integrations/connectors/source-faker/icon.svg" width="40" height="40"> source-faker test report (commit [`26884cd0db`](https://github.com/airbytehq/airbyte/pull/27957/commits/26884cd0dbe194727b5982e7b67c96e7b60573a6)) - ❌

⏲️  Total pipeline duration: 03mn30s 

| Step                                                                | Result   |
|:--------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-faker/metadata.yaml | ✅        |
| Connector version semver check                                      | ✅        |
| Connector version increment check                                   | ❌        |
| QA checks                                                           | ✅        |
| Code format checks                                                  | ✅        |
| Connector package install                                           | ✅        |
| Build source-faker docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                          | ✅        |
| Acceptance tests                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/bnchrch_ci_poetry-pipelines/1690249379/26884cd0dbe194727b5982e7b67c96e7b60573a6/source-faker/4.0.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-faker test
```

bnchrch commented on airbyte-ci/connectors/ci_code_validator/__init__.py:
@airbytehq/connector-operations Gut check: We should remove this.

There is
1. no use of `ci_sonar_qube` in code
2. no use of `ci_changes_detection` in code 
3. the is one use of `ci_sonar_qube` in a github action: https://github.com/airbytehq/airbyte/pull/27957/files#diff-c46f6e02d0050829e4e9f4a4bd0eb29da35853e0786b7ddf348f665085c561da
4. But that action is not used.

erohmensing replied:
I've never heard of this, if that helps anything!

bnchrch commented on .github/workflows/test-performance-command.yml:
This should be deleted. 

See: https://github.com/airbytehq/airbyte/pull/27957/files#r1272911875

bnchrch commented:
@airbytehq/connector-operations This is ready for review!

Most of this is
1. a folder restructure
2. a tooling change
3. a python version bump

The only file with net new logic is how we install pipx and use it in the pipelines here

[...tor_ops/pipelines/actions/environments.py → ...pelines/pipelines/actions/environments.py](https://github.com/airbytehq/airbyte/pull/27957/files#diff-b1c73385b3a55c0d61d11e4323076b841402c0401100fc95d902b9c4d87609fe)

Please let me know what you think, I would like this to be a speedy merge to avoid conflicts!

erohmensing reviewed (approved):
Installation was a breeze! Messed around with formatting a source, testing the orchestrator and metadata lib. I'm still not sure (via READMEs alone, and after a quick non-deep perusal of CI changes) how to run tests for the pipelines. 

Some non-blocking feedback - I think we should merge this improvement before trying to tweak anything much more: 
* I was able to run the QA engine from where it lives with the instructions provided. I find it sort of weird it lives under airbyte ci but has a separate installation and run process
* It looks like the README of airbyte-ci lives under `airbyte-ci/connectors/pipelines` - i'd expect it to live under `airbyte-ci` or maybe `airbyte-ci/connectors`. I guess this makes sense because of the install command being `pipx install airbyte-ci/connectors/pipelines`... this installation path confuses me since i expect airbyte-ci to be able to run the metadata service as well (and the qa engine as described above)

bnchrch commented on airbyte-ci/connectors/pipelines/pyproject.toml:
```diff
@@ -0,0 +1,27 @@
+
+[tool.poetry.dependencies]
+python = "^3.10"
+dagger-io = "^0.6.4"
```
As I understand it. This is expected!

https://github.com/pypa/pipx#how-is-it-different-from-pip

pipx actually installs all dependencies for the specified package in a directory isolated from your current python environment.

The packages you have in you global env or pyenv will have no effect on the package dependencies install with pipx and vice versa

So it makes sense to me that the dagger-io version in your pyenv shell didnt change.

But can you tell me what you ran to get the version for dagger-io?

bnchrch commented on airbyte-ci/connectors/pipelines/pipelines/consts.py:
```diff
@@ -32,6 +32,6 @@
 GRADLE_BUILD_CACHE_PATH = f"{GRADLE_CACHE_PATH}/build-cache-1"
 GRADLE_READ_ONLY_DEPENDENCY_CACHE_PATH = "/root/gradle_dependency_cache"
-LOCAL_REPORTS_PATH_ROOT = "tools/ci_connector_ops/pipeline_reports/"
+LOCAL_REPORTS_PATH_ROOT = "airbyte-ci/connectors/pipelines/pipeline_reports/"
```
Ooh good catch. Let me see whats going on

bnchrch replied:
Ah I see whats going on

**empty pipeline_reports folder created**
This is an is issue with where we [create the report directory](https://github.com/airbytehq/airbyte/blob/370c3fb884ee3ae47684c9d595169d23ffd08037/airbyte-ci/connectors/pipelines/pipelines/utils.py#L452) which is only discovered if you run the `airbyte-ci` cmd not in the root.

Ill move this closer to us writing the log

**And the actual test reports are written a very nested location:**
It is very nested but expected

The log path `airbyte-ci/connectors/pipelines/pipeline_reports/airbyte-ci/connectors/test/manual/bnchrch_ci_poetry-pipelines/1690300788/26884cd0dbe194727b5982e7b67c96e7b60573a6/dagger.log` has three discrete parts

1. The parent folder `airbyte-ci/connectors/pipelines/`
2. The airbyte-ci defined path structure `airbyte-ci/connectors/test/`
3. the git info `manual/bnchrch_ci_poetry-pipelines/1690300788/26884cd0dbe194727b5982e7b67c96e7b60573a6`

so you will note that #1 and #2 are nearly identical. 

and that turns out is a coincidence.

The airbyte-ci defined path structure is derived from the click command hierarchy `airbyte-ci connectors test --name=...`

bnchrch commented on airbyte-ci/connectors/pipelines/README.md:
````diff
@@ -1,30 +1,50 @@
-N.B: This project will eventually be moved to `airbyte-ci` root directory.
 
+### Installation for development
+If you are developing on pipelines, we recommend installing airbyte-ci in editable mode:
+
+```bash
+pipx install airbyte-ci/connectors/pipelines/ --force --editable
+```
````
Thumbs up to using poetry for dev!

Seems a lot more approachable and intuitive :)

bnchrch commented on airbyte-ci/connectors/pipelines/README.md:
````diff
@@ -1,30 +1,50 @@
+pipx install airbyte-ci/connectors/pipelines/
 ```
 
+This command installs `airbyte-ci` and makes it globally available in your terminal.
````
Could you elaborate a bit more on the why and what an ideal install instruction looks like?

alafanechere replied:
I think that after you run pipx install the `airbyte-ci` the CLI is not directly available in the PATH.
I suggest to add something like: `Open a new terminal session or source your bash profile to make `airbyte-ci` available in your path.`

octavia-squidington-iii commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/68758f99ba89dfbd385b70129f33cd89762aee48/airbyte-integrations/connectors/source-faker/icon.svg" width="40" height="40"> source-faker test report (commit [`68758f99ba`](https://github.com/airbytehq/airbyte/pull/27957/commits/68758f99ba89dfbd385b70129f33cd89762aee48)) - ❌

⏲️  Total pipeline duration: 03mn22s 

| Step                                                                | Result   |
|:--------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-faker/metadata.yaml | ✅        |
| Connector version semver check                                      | ✅        |
| Connector version increment check                                   | ❌        |
| QA checks                                                           | ✅        |
| Code format checks                                                  | ✅        |
| Connector package install                                           | ✅        |
| Build source-faker docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                          | ✅        |
| Acceptance tests                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/bnchrch_ci_poetry-pipelines/1690335663/68758f99ba89dfbd385b70129f33cd89762aee48/source-faker/4.0.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-faker test
```

bnchrch commented:
@alafanechere @pedroslopez 

Just updated the branch to
1. Have poetry based dev install instructions
2. Have pipeline test deps declared
3. Remove ci_code_validator folder and related action
4. Upgrade `.github/workflows/connector-performance-command.yml` to 3.10
5. Ran the python formatter on airbyte-ci
6. Updated the logging code to not create the dagger.log parent folder until nessesary

Im going to run some test pipelines now to ensure everything works as planned

octavia-squidington-iii commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/4090391bacd54a11de7059de732f1630349361ba/airbyte-integrations/connectors/source-faker/icon.svg" width="40" height="40"> source-faker test report (commit [`4090391bac`](https://github.com/airbytehq/airbyte/pull/27957/commits/4090391bacd54a11de7059de732f1630349361ba)) - ❌

⏲️  Total pipeline duration: 03mn17s 

| Step                                                                | Result   |
|:--------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-faker/metadata.yaml | ✅        |
| Connector version semver check                                      | ✅        |
| Connector version increment check                                   | ❌        |
| QA checks                                                           | ✅        |
| Code format checks                                                  | ✅        |
| Connector package install                                           | ✅        |
| Build source-faker docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                          | ✅        |
| Acceptance tests                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/bnchrch_ci_poetry-pipelines/1690336063/4090391bacd54a11de7059de732f1630349361ba/source-faker/4.0.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-faker test
```

bnchrch commented:
**Test**
- [GHA run ](https://github.com/airbytehq/airbyte/actions/runs/5663728190)with args `--name=source-postgres --name=destination-postgres --name=source-fresh-desk` 

**prepublish**
- [source-postgres](https://github.com/airbytehq/airbyte/actions/runs/5663736440)
- [source-fresh-desk](https://github.com/airbytehq/airbyte/actions/runs/5663742866)

alafanechere commented:
@bnchrch I pushed a gitignore change to ignore the legacy pipeline reports paths. Otherwise a `git add .` will stage all the old report file a developer might have locally.

alafanechere commented:
@bnchrch I updated the poetry lock following the test dependency declaration.
